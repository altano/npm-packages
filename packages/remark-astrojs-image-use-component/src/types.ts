import { is } from "unist-util-is";

import type { Node, Data } from "unist";
import type { MdxJsxFlowElement } from "mdast-util-mdx-jsx";
import type { Image as MarkdownImage } from "mdast";

export interface MdxJsxImageElement extends MdxJsxFlowElement {
  type: "mdxJsxFlowElement";
  name: "img";
}

export interface MdxJsxPictureElement extends MdxJsxFlowElement {
  type: "mdxJsxFlowElement";
  name: "picture";
}

export function isMdxJsxImageElement(
  node: Node<Data>,
): node is MdxJsxImageElement {
  return is(node, { type: "mdxJsxFlowElement", name: "img" });
}

export function isMdxJsxPictureElement(
  node: Node<Data>,
): node is MdxJsxPictureElement {
  return is(node, { type: "mdxJsxFlowElement", name: "picture" });
}

export function isMarkdownImage(node: Node<Data>): node is MarkdownImage {
  return is(node, { type: "image" });
}
