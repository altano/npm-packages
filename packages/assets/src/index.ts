import { createRequire } from "node:module";
import { readFile } from "node:fs/promises";

const resolve = createRequire(import.meta.url).resolve;

/**
 * @returns the file path for the Inter woff font file at the given weight
 */
export function getInterPath(fontWeight: number): string {
  return resolve(
    `@fontsource/inter/files/inter-latin-${fontWeight}-normal.woff`,
  );
}

/**
 * @returns the node Buffer for the Inter woff font file at the given weight
 */
export async function getInterBuffer(fontWeight: number): Promise<Buffer> {
  return readFile(getInterPath(fontWeight));
}
