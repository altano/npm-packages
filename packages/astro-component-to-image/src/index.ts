import satori from "satori";
import { defineMiddleware } from "astro/middleware";
import sharp, { type Sharp, type FormatEnum as SharpFormatEnum } from "sharp";
import { html as htmlToVNode } from "satori-html";
import { contentType } from "mime-types";
import he from "he";

import type { MiddlewareResponseHandler, APIContext } from "astro";
import type { SatoriOptions as SatoriOptionsOrig } from "satori";

export type SatoriOptions = SatoriOptionsOrig;
export type SharpOptions = NonNullable<Parameters<Sharp["toFormat"]>[1]>;
export type SharpFormats = keyof SharpFormatEnum;

// TODO Needs lots of tests

export type ImageOptions<Format extends SharpFormats> = {
  format: Format;
  options?: SharpOptions;
};

type Filename<Format extends SharpFormats> = `${string}.${Format}`;

export type Options<Format extends SharpFormats> = {
  /**
   * Any output format that the sharp library accepts, as a string.
   * e.g. "avif", "jpg", "png", "webp", "gif", etc.
   *
   * API documentation: https://sharp.pixelplumbing.com/api-output
   */
  format: Format;
  /**
   * Options that the vercel/satori library accepts. At the very least, you must
   * specify dimensions and one font to use.
   *
   * API documentation:
   * https://github.com/vercel/satori/blob/0a258931fe2291bdd461103780ac01e3c700b845/src/satori.ts#L18-L40
   */
  getSatoriOptions(
    context: APIContext,
    response: Response,
    format: Format,
  ): Promise<SatoriOptions>;
  /**
   * Any output options that the sharp library accepts, e.g.:
   *
   *   { quality: 50 }
   *
   * API documentation: https://sharp.pixelplumbing.com/api-output#toformat
   */
  getSharpOptions?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<SharpOptions>;
  /**
   * This function must return true for the given request or the route will not
   * be converted to an image. By default, only components/endpoints that return
   * a text/html Response and have a route ending in "<name>.<format>", e.g.
   * image.png, will be converted to an image. Use this method to override that
   * behavior and provide your own heuristic.
   */
  shouldReplace?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<boolean>;
  /**
   * The default filename for a component such as `image.png.astro` is
   * `image.png`. You may use this function to override that default. Only
   * relevant in SSG: in SSR there is no filename and Astro just responds with
   * the file image contents.
   */
  getFilename?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<Filename<Format>>;
};

function getContentType<Format extends SharpFormats>(format: Format): string {
  const type = contentType(format);
  return type === false ? `image/${format}` : type;
}

async function defaultGetFilename<Format extends SharpFormats>(
  format: Format,
  context: APIContext,
): Promise<string> {
  const requestUrl = context.url.toString();
  new RegExp(String.raw`\.${format}\/?$`).test(requestUrl);

  const basename = context.url.pathname.split("/").filter(Boolean).at(-1);

  if (basename == null) {
    throw new Error(
      `Could not find basename of URL: ${context.url.toString()}`,
    );
  }

  return basename;
}

async function defaultGetSharpOptions(): Promise<SharpOptions> {
  return { quality: 100 };
}

async function defaultShouldReplace<Format extends SharpFormats>(
  context: APIContext,
  response: Response,
  format: Format,
): Promise<boolean> {
  const contentType = response.headers?.get("content-type");

  // In SSG there will be a trailing slash in the request URL. In SSR/dev a
  // user might type it in without the slash. Handle both.
  const requestUrl = context.url.toString();
  const isImageOfType = new RegExp(String.raw`\.${format}\/?$`).test(
    requestUrl,
  );

  return (
    isImageOfType &&
    // Don't rewrite other things like Endpoints. Only interested in components.
    response instanceof Response &&
    // Don't rewrite anything that isn't html.
    contentType === "text/html"
  );
}

export function createHtmlToImageMiddleware<Format extends SharpFormats>({
  format,
  getSharpOptions,
  shouldReplace,
  getFilename,
  getSatoriOptions,
}: Options<Format>): MiddlewareResponseHandler {
  return defineMiddleware(async (context, next) => {
    const response = await next();
    const replace =
      shouldReplace == null
        ? await defaultShouldReplace(context, response, format)
        : await shouldReplace(context, response, format);

    if (!replace) {
      // This request isn't relevant to us. Don't modify the response.

      // TODO Remove when issue is fixed in astro
      // https://github.com/withastro/astro/issues/8045
      if (isEndpointOutput(response)) {
        return new Response(response.body, {
          status: 200,
          headers: response.headers,
        });
      }

      return response;
    }

    const [responseText, satoriOptions, sharpOptions, filename] =
      await Promise.all([
        response.text(),
        getSatoriOptions(context, response, format),
        getSharpOptions == null
          ? defaultGetSharpOptions()
          : getSharpOptions(context, response, format),
        getFilename == null
          ? defaultGetFilename(format, context)
          : getFilename(context, response, format),
      ]);

    // html text => vnode
    const responseTextWithDecodedHtmlEntities = he.decode(responseText);
    const vnode = htmlToVNode(responseTextWithDecodedHtmlEntities);

    // vnode => svg
    const svg = await satori(vnode as React.ReactNode, satoriOptions);

    // svg => image
    const imageBuffer = await sharp(Buffer.from(svg))
      .toFormat(format, sharpOptions)
      .toBuffer();

    return new Response(imageBuffer, {
      headers: {
        "Content-Type": getContentType(format),
        "Content-Disposition": `inline; filename="${filename}"`,
      },
    });
  });
}

function isEndpointOutput(endpointResult: unknown): boolean {
  return (
    endpointResult != null &&
    !(endpointResult instanceof Response) &&
    typeof endpointResult === "object" &&
    "body" in endpointResult &&
    typeof endpointResult.body === "string"
  );
}
