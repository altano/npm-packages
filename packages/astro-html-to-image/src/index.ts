import satori from "satori";
import { defineMiddleware } from "astro/middleware";
import { Resvg } from "@resvg/resvg-wasm";
import { html as htmlToVNode } from "satori-html";
import { contentType } from "mime-types";
import he from "he";

import type { MiddlewareResponseHandler, APIContext } from "astro";
import type { SatoriOptions } from "satori";
import type { ResvgRenderOptions } from "@resvg/resvg-wasm";

export type SvgOptions = SatoriOptions;
export type ImageRenderOptions = ResvgRenderOptions;
export type ImageFormat = "png";

// TODO Needs lots of tests

export type ImageOptions<Format extends ImageFormat> = {
  format: Format;
  options?: ImageRenderOptions;
};

type Filename<Format extends ImageFormat> = `${string}.${Format}`;

export type Runtime = "nodejs";

export async function initialize({
  runtime,
}: {
  runtime: Runtime;
}): Promise<void> {
  switch (runtime) {
    case "nodejs":
      {
        const initialize = await import("./initializeResvg-Nodejs");
        await initialize.default();
      }
      break;
    default:
      runtime satisfies never;
  }
}

export type Options<Format extends ImageFormat> = {
  /**
   * Your JavaScript runtime, e.g. "nodejs"
   */
  runtime: Runtime;
  /**
   * Any output format that @resvg/resvg-wasm library accepts, as a string.
   * Currently only "png"
   */
  format: Format;
  /**
   * Options that the vercel/satori library accepts. At the very least, you must
   * specify dimensions and one font to use.
   *
   * API documentation:
   * https://github.com/vercel/satori/blob/0a258931fe2291bdd461103780ac01e3c700b845/src/satori.ts#L18-L40
   */
  getSvgOptions(
    context: APIContext,
    response: Response,
    format: Format,
  ): Promise<SvgOptions>;
  /**
   * Any output options that the @resvg/resvg-js library accepts, e.g.:
   *
   *   { background: "rgba(255,255,255, 0.8)" }
   *
   * API documentation: https://github.com/yisibl/resvg-js/blob/main/index.d.ts#L3
   */
  getImageOptions?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<ImageRenderOptions>;
  /**
   * This function must return true for the given request or the route will not
   * be converted to an image. By default, only components/endpoints that return
   * a text/html Response and have a route ending in "<name>.<format>", e.g.
   * image.png, will be converted to an image. Use this method to override that
   * behavior and provide your own heuristic.
   */
  shouldReplace?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<boolean>;
  /**
   * The default filename for a component such as `image.png.astro` is
   * `image.png`. You may use this function to override that default. Only
   * relevant in SSG: in SSR there is no filename and Astro just responds with
   * the file image contents.
   */
  getFilename?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<Filename<Format>>;
};

function getContentType<Format extends ImageFormat>(format: Format): string {
  const type = contentType(format);
  return type === false ? `image/${format}` : type;
}

export async function defaultGetFilename<Format extends ImageFormat>(
  format: Format,
  context: APIContext,
): Promise<string> {
  const requestUrl = context.url.toString();
  new RegExp(String.raw`\.${format}\/?$`).test(requestUrl);

  const basename = context.url.pathname.split("/").filter(Boolean).at(-1);

  if (basename == null) {
    throw new Error(
      `Could not find basename of URL: ${context.url.toString()}`,
    );
  }

  return basename;
}

export async function defaultGetImageOptions(): Promise<ImageRenderOptions> {
  return {
    textRendering: 1, // optimizeLegibility
    imageRendering: 0, // optimizeQuality
  };
}

export async function defaultShouldReplace<Format extends ImageFormat>(
  context: APIContext,
  response: Response,
  format: Format,
): Promise<boolean> {
  const contentType = response.headers?.get("content-type");

  // In SSG there will be a trailing slash in the request URL. In SSR/dev a
  // user might type it in without the slash. Handle both.
  const requestUrl = context.url.toString();
  const isImageOfType = new RegExp(String.raw`\.${format}\/?$`).test(
    requestUrl,
  );

  return (
    isImageOfType &&
    // Don't rewrite other things like Endpoints. Only interested in components.
    response instanceof Response &&
    // Don't rewrite anything that isn't html.
    contentType === "text/html"
  );
}

let isWasmInitialized = false;

export function createHtmlToImageMiddleware<Format extends ImageFormat>({
  runtime,
  format,
  getImageOptions,
  shouldReplace,
  getFilename,
  getSvgOptions,
}: Options<Format>): MiddlewareResponseHandler {
  return defineMiddleware(async (context, next) => {
    if (!isWasmInitialized) {
      isWasmInitialized = true;
      try {
        await initialize({ runtime });
      } catch (err) {
        // Restarting the dev server causes us to unnecessarily initialize the
        // wasm module twice. Let's just ignore the error (since we can't
        // prevent it).
        const shouldIgnore =
          err instanceof Error &&
          err.message?.startsWith("Already initialized");
        if (!shouldIgnore) {
          // re-throw all other errors
          throw err;
        }
      }
    }

    const response = await next();
    const replace =
      shouldReplace == null
        ? await defaultShouldReplace(context, response, format)
        : await shouldReplace(context, response, format);

    if (!replace) {
      // This request isn't relevant to us. Don't modify the response.

      // TODO Remove when issue is fixed in astro
      // https://github.com/withastro/astro/issues/8045
      if (isEndpointOutput(response)) {
        return new Response(response.body, {
          status: 200,
          headers: response.headers,
        });
      }

      return response;
    }

    const [responseText, svgOptions, imageOptions, filename] =
      await Promise.all([
        response.text(),
        getSvgOptions(context, response, format),
        getImageOptions == null
          ? defaultGetImageOptions()
          : getImageOptions(context, response, format),
        getFilename == null
          ? defaultGetFilename(format, context)
          : getFilename(context, response, format),
      ]);

    // html text => vnode
    const responseTextWithDecodedHtmlEntities = he.decode(responseText);
    const vnode = htmlToVNode(responseTextWithDecodedHtmlEntities);

    // vnode => svg
    const svg = await satori(vnode as React.ReactNode, svgOptions);

    // svg => image
    const imageResvgArr = new Resvg(svg, imageOptions).render().asPng();
    const imageBuffer = Buffer.from(imageResvgArr);

    return new Response(imageBuffer, {
      headers: {
        "Content-Type": getContentType(format),
        "Content-Disposition": `inline; filename="${filename}"`,
      },
    });
  });
}

function isEndpointOutput(endpointResult: unknown): boolean {
  return (
    endpointResult != null &&
    !(endpointResult instanceof Response) &&
    typeof endpointResult === "object" &&
    "body" in endpointResult &&
    typeof endpointResult.body === "string"
  );
}
