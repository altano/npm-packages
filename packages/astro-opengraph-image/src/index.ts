import { createImageMiddleware } from "./createImageMiddleware";

export const createOpenGraphImageMiddleware = createImageMiddleware;
