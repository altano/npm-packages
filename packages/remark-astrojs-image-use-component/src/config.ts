export interface RemarkAstroJSImageUseComponentOptions {
  /**
   * Convert ![]() style markdown images to AstroJS' Image component
   *
   * @default true
   */
  convertMarkdownImages?: boolean;
  /**
   * Convert <img> jsx images to AstroJS' Image component
   *
   * @default true
   */
  convertJsxImages?: boolean;
  /**
   * Convert <picture> jsx pictures to AstroJS' Picture component
   *
   * @default true
   */
  convertJsxPictures?: boolean;
}

export type RemarkAstroJSImageUseComponentConfig =
  Required<RemarkAstroJSImageUseComponentOptions>;

const configDefaults: RemarkAstroJSImageUseComponentConfig = {
  convertMarkdownImages: true,
  convertJsxImages: true,
  convertJsxPictures: true,
};

// import type { Element } from "hast";
export function getConfig(
  options: RemarkAstroJSImageUseComponentOptions,
): RemarkAstroJSImageUseComponentConfig {
  return Object.assign(configDefaults, options);
}
