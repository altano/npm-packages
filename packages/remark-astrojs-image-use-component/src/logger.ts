import { Logger } from "@altano/remark-plugin-helpers";

const logger = new Logger("@altano/remark-astrojs-image-use-component");

export default logger;
