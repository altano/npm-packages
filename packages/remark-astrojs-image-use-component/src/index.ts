import { getConfig } from "./config";
import { createAstroJSImageElement } from "./mdx/createAstroJSImageElement";
import {
  isMarkdownImage,
  isMdxJsxImageElement,
  isMdxJsxPictureElement,
} from "./types";
import { visitAndReplace } from "./mdx/visitAndReplace";
import { ensureMdxjsEsmExists } from "./mdx/mdxjsEsm";

import type { Plugin } from "unified";
import type { Node, Data } from "unist";
import type { Root } from "mdast";
import type { RemarkAstroJSImageUseComponentOptions } from "./config";

/**
 * A Remark plugin for converting images in mdx to @astrojs/image components
 */
const remarkAstroJSImageUseComponentPlugin: Plugin<
  [RemarkAstroJSImageUseComponentOptions?],
  Root
> = (options) => (tree: Root) => {
  // console.log(`remarkAstroJSImageUseComponentPlugin running on tree: `, tree);

  const config = getConfig(options ?? {});
  let needToAddImageImport = false;
  let needToAddPictureImport = false;

  if (config.convertMarkdownImages) {
    visitAndReplace(tree, "image", (node: Node<Data>) => {
      if (!isMarkdownImage(node)) {
        return;
      }
      needToAddImageImport = true;
      return createAstroJSImageElement(node);
    });
  }

  if (config.convertJsxImages) {
    visitAndReplace(tree, undefined, (node: Node<Data>) => {
      if (!isMdxJsxImageElement(node)) {
        return;
      }
      needToAddImageImport = true;
      return createAstroJSImageElement(node);
    });
  }

  if (config.convertJsxPictures) {
    visitAndReplace(tree, undefined, (node: Node<Data>) => {
      if (!isMdxJsxPictureElement(node)) {
        return;
      }
      needToAddPictureImport = true;
      return createAstroJSImageElement(node);
    });
  }

  if (needToAddImageImport) {
    ensureMdxjsEsmExists(tree, "Image", "@astrojs/image/components");
  }

  if (needToAddPictureImport) {
    ensureMdxjsEsmExists(tree, "Picture", "@astrojs/image/components");
  }
};

export default remarkAstroJSImageUseComponentPlugin;
