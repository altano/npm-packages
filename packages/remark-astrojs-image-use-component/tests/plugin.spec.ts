import plugin from "../src/index";
import { describe } from "vitest";
import { testByFixtures } from "@altano/remark-plugin-test-util/src/testByFixture";

describe("remark-astrojs-image-use-component", async () => {
  await testByFixtures(plugin);
});
