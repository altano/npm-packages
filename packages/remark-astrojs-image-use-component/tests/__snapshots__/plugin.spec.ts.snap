// Vitest Snapshot v1

exports[`remark-astrojs-image-use-component > [test #0] data-component-ignore 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
function _createMdxContent(props) {
  return (
    <>
      <img src="jsx-image.png" data-component-ignore />
      {"\\n"}
      <picture src="jsx-picture.png" data-component-ignore />
    </>
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-use-component > [test #1] existing-import 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
import { Image } from "@astrojs/image/components";
function _createMdxContent(props) {
  return (
    <>
      <Image src="jsx-image.png" />
      {"\\n"}
      <Image src="astro-image.png" />
    </>
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-use-component > [test #2] jsx-image 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
import { Image } from "@astrojs/image/components";
function _createMdxContent(props) {
  return (
    <>
      <Image src="jsx-image.png" />
      {"\\n"}
      <Image src="jsx-image.png" width={500} height={500} />
      {"\\n"}
      <Image src="jsx-image.png" alt="Alt Text" />
      {"\\n"}
      <Image src="jsx-image.png" title="Title Text" />
    </>
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-use-component > [test #3] jsx-picture 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
import { Picture } from "@astrojs/image/components";
import { Image } from "@astrojs/image/components";
function _createMdxContent(props) {
  return (
    <>
      <Picture src="jsx-picture.png" />
      {"\\n"}
      <Picture src="jsx-picture.png" width={500} height={500} />
      {"\\n"}
      <Picture src="jsx-picture.png" alt="Alt Text" />
      {"\\n"}
      <Picture src="jsx-picture.png" title="Title Text" />
      {"\\n"}
      <Picture>
        <source srcset="jsx-picture-2x.png" media="(min-width: 800px)" />
        <Image src="jsx-picture-1x.png" alt="alt" />
      </Picture>
    </>
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-use-component > [test #4] markdown 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
import { Image } from "@astrojs/image/components";
function _createMdxContent(props) {
  const _components = Object.assign(
    {
      p: "p",
    },
    props.components,
  );
  return (
    <>
      <_components.p>
        <Image src="markdown.png" alt="" />
      </_components.p>
      {"\\n"}
      <_components.p>
        <Image src="markdown.png" alt="Alt Text" />
      </_components.p>
      {"\\n"}
      <_components.p>
        <Image src="markdown.png" alt="Alt Text" title="Title Text" />
      </_components.p>
    </>
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;
